Statement optimizations

tinwelint opened the pull request:
A couple of commits to improve some aspects of transaction handling and reading records.

tinwelint committed:
More efficient reuse of statement instances

Each KernelStatement instance is coupled with a KTI instances
and follows same initialize() calls as it. Likewise each StoreStatement
instance is coupled with a KernelStatement instance.

Starting a new transaction will extremely rarely create a new KTI instance,
otherwise most times pick one from a ThreadLocal with all these instances
already created, just waiting for a call to initialize() to be ready for use.
Likewise starting a statement just calls KernelStatement#acquire(), which in
turn calls StorageStatement#acquire(). Statement state is still cleated in
KernelStatement/StorageStatement#close(), but same instances will be reused
for the new statement (this was already done for KernelStatement). This commit
removes external pooling of StorageStatement instances since it's no longer
needed, so it's a net simplification of the code.

tinwelint committed:
More efficient nodeExists

by using CommonAbstractStore#isInUse instead of reading the whole record
to figure that out.

tinwelint committed:
Ensures correct calculation w/ paranthesis

tinwelint committed:
More efficient getRelationshipById

by evaluating and keeping all the fields in the read relationship record,
handing them to the RelationshipProxy at construction instead of only
handing over the id and let the next interaction with that relationship
require to read the relationship record again.

tinwelint committed:
Flatten method call stack in CommonAbstractStore#getRecord

Instead favoring composition so that the JVM can inline things.
Flattening call stack from 4-5 down to 2. Apparently this can make
a big difference up to 20-30% for access patterns that are very heavy
on getRecord.

tinwelint committed:
Flatten RecordFormat call stacks

and favoring composition of small methods for ability to inline instead.

tinwelint committed:
More efficient getRecordSize()

which allows getting record size to be inlined in CommonAbstractStore.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1089,11 +1038,36 @@ public boolean next()
+                        record.setId( currentId );
+                        long pageId = pageIdForRecord( currentId );
+                        if ( pageId == pageCursor.getCurrentPageId() ||
+                                pageCursor.next( pageIdForRecord( currentId ) ) )
```
No need to recompile pageId for the next call.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -1089,11 +1038,36 @@ public boolean next()
+                        if ( pageId == pageCursor.getCurrentPageId() ||
+                                pageCursor.next( pageIdForRecord( currentId ) ) )
+                        {
+                            int offset = offsetForId( currentId );
```
If you place the calls to offsetForId and pageIdForRecord next to each other, and both inline, then the JIT can extract both the modulo for the offset and the division for the page id, from the same divide instruction.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java:
```diff
@@ -35,14 +34,14 @@ public void shouldReleaseItselfWhenClosed() throws Exception
 
         // then
-        verify( transaction ).releaseStatement( statement );
+//        verify( transaction ).releaseStatement( statement );
```
Looks like this should be fixed up

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/StatementLifecycleTest.java:
```diff
@@ -60,6 +59,6 @@ public void shouldReleaseWhenAllNestedStatementsClosed() throws Exception
 
         // then
-        verify( transaction ).releaseStatement( statement );
+//        verify( transaction ).releaseStatement( statement );
```
And here

chrisvest commented:
Interesting stuff.

chrisvest commented:
@tinwelint Really weird test failures; many about enabling upgrading.

tinwelint commented:
I'll rebase once https://github.com/neo4j/neo4j/pull/6495 is in and fix the issue, because it will be much simpler then.

tinwelint committed:
Uses KTI storage statement during commit

instead of creating new ones, many during each commit

systay commented:
@tinwelint #6495 has been merged

tinwelint commented:
@systay thanks for notice

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelStatement.java:
```diff
@@ -130,23 +151,22 @@ void assertOpen()
-
-        return (referenceCount == 0);
+        return (--referenceCount == 0);
     }
```
The `acquire` and `release` methods are not symmetric with regards to their visibility, or what they do to the `storeStatement`.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -169,6 +168,8 @@ public KernelTransactionImplementation( StatementOperationParts operations,
         this.pool = pool;
         this.clock = clock;
         this.tracer = tracer;
+        this.storageStatement = storeLayer.acquireStatement();
```
I'm not seeing where this guy is closed.

tinwelint replied:
Since that method no longer acquires a statement, but merely creates one it should be renamed to `newStatement()`. Calling `acquire()` acquires it. It's quite internal and KTI is basically the only user so that slight inconvenience of this change is minimal btw. I'll do this rename.

... and so that guy shouldn't be closed by KTI, but will actually will be by KernelStatement.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java:
```diff
@@ -283,13 +283,16 @@ public Relationship getRelationshipById( long id )
         }
```
These ID bounds checks should be removed and left to the `StorageEngine`, so we can remove the `MAX_RELATIONSHIP_ID` and `MAX_NODE_ID` constants. The constants in `IdType` that feed our `IdGenerators` should also be removed, and we should source that data from the `RecordFormat` somehow.

tinwelint replied:
Right, that's true. Shall we leave that to another PR perhaps?

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -227,7 +231,7 @@ protected void loadStorage()
         {
-            // TODO: Just throw IOException, add proper handling further up
+            // Just throw IOException, add proper handling further up
             throw new UnderlyingStorageException( e );
```
This comment is now wrong.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/CommonAbstractStore.java:
```diff
@@ -257,34 +261,35 @@ private void extractHeaderRecord() throws IOException
                 do
                 {
-                    pageCursor.setOffset( (int) (id % getRecordsPerPage() * getRecordSize()) );
+                    pageCursor.setOffset( (int) (id % getRecordsPerPage() * recordSize) );
```
We could also add a parenthesis here.

tinwelint replied:
We could use the `RecordPageLocationCalculator` here as well.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/format/BaseOneByteHeaderRecordFormat.java:
```diff
@@ -38,72 +34,32 @@
-     */
-    protected abstract void doWrite( RECORD record, PageCursor cursor, int recordSize, PagedFile storeFile )
-            throws IOException;
+    protected boolean isInUse( byte firstByte )
```
I think we should have some javadocs on these methods.

chrisvest commented on community/kernel/src/main/java/org/neo4j/storageengine/api/StorageStatement.java:
```diff
@@ -95,6 +95,8 @@
      */
     Cursor<RelationshipItem> relationshipsGetAllCursor();
 
+    void acquire();
```
Please add javadocs that describe what this is supposed to mean.

chrisvest commented:
@tinwelint I've done my first review.

tinwelint committed:
Uses RecordPageLocationCalculator in more places

tinwelint committed:
Documentation on BaseOneByteHeaderRecordFormat utility methods

tinwelint committed:
Clarifies StorageStatement lifecycle

by adding more documenting to class and new methods, but also with
better method naming.

tinwelint commented:
@chrisvest comments addressed.

chrisvest commented:
@tinwelint Nice, but now it needs a rebase.
